src/book.py

```diff
@@ -58,9 +58,11 @@ def set_price(self, price: float) -> None:
         self.__price = price
 
     def __eq__(self, other):
-        if isinstance(other, Book):
-            return other.__title == self.__title
-        return other in self.__title
+        return other.__title == self.__title
```

Comment:
Acho que me entendeu errado, eu gostei do `isinstance`, eu só não entendi o `other in self.__title`. Simplificando, chuto que `isinstance && ==` em uma única expressão já resolve.

Reply:
```
    def __eq__(self, other):
        if isinstance(other, Book):
            return other.__title == self.__title
        return other in self.__title
```
Ata, bom, a ideia aqui é que se você tentar verificar se um objeto do tipo **Book** é igual a outro do tipo **Book**, a regra de negocio determina que ele são iguais se tiverem o mesmo titulo, então eu acesso o atributo e verifico se são iguais.
```
def search(self, text_search: str) -> List:
        if len(text_search) < 2:
            raise Exception("Need to have at least two characters")
        if text_search not in self.__lista:
            raise Exception(text_search + " not found")
        matching = [s for s in self.__lista if text_search == s]
        return matching
```
Porem quando fui utilizar o **in**, que itera pela lista de livro, e verificar se um objeto do tipo **Book** é igual a um tipo **str**, isso gerava um erro já que o objeto do tipo **str** não tem o atributo **.__title**, então fui obrigado a verificar se ele é do tipo **Book** para acessar o atributo e se não for eu assumia que era uma string e via se essa string estava contida dentro do titulo.

Reply:
```
    def search(self, text_search: str) -> List:
        if len(text_search) < 2:
            raise Exception("Need to have at least two characters")

        matching: List[Book] = []
        matching = [book for book in self.__lista if text_search in book.title]
        return matching
```
Gostei mais desse jeito, onde fica mais claro que estou olhando para o titulo e verificando se a string está dentro dele.

Reply:
EU quis dizer isso aqui:

`
def __eq__(self, other):
        if isinstance(other, Book):
            return other.__title == self.__title
        return false
`
Se é o mesmo tipo, você compara o título, se não, false direto. A comparação do título vai dizer em segundo nível, se é o mesmo livro ou não. Aqui vale a pena pensar, se não forem do mesmo tipo, não vale a pena nem comparar mais nada.

Reply:
Então, mas é que a minha ideia inicial é que o `def __eq__(self, other)` fosse capaz de fazer duas coisas.

1. Verificar se dois livros são igual, olhando para o atributo titulo deles.

Reply:
O problema do caso 2, é que foge um pouco da questão da tipagem da orientação a objetos, se eu posso comparar tipos diferentes, por que existe a restrição de ser uma string? Pode deixar confuso o entendimento do software. Eu acho arriscada essa flexibilidade.

Reply:
Concordo, já troquei a logica, e agora a classe CollectionBooks ficou com a responsabilidade de descobrir se a String passada, text_search, tem relação com algum titulo de livro que ela guarda.
```
  def search(self, text_search: str) -> List:
        if len(text_search) < 2:
            raise Exception("Need to have at least two characters")
        matching: List[Book] = []
        matching = [book for book in self.__lista if text_search in book.title]
        return matching
```